ftl/python/builder.py

```diff
@@ -43,60 +43,77 @@ def __init__(self, ctx, args, cache_version_str):
+                    key = req_txt_builder.GetCacheKey()
+                    cached_req_txt_img = self._cache.Get(key)
+            if cached_req_txt_img is not None:
+                req_txt_builder.SetImage(cached_req_txt_img)
+            else:
+                with ftl_util.Timing("installing pip packages"):
+                    pkg_descriptor = ftl_util.descriptor_parser(
+                        self._descriptor_files, self._ctx)
+                    self._pip_install(pkg_descriptor)
+
+                with ftl_util.Timing("resolving whl paths"):
+                    whls = self._resolve_whls()
+                    pkg_dirs = [self._whl_to_fslayer(whl) for whl in whls]
+
+                req_txt_imgs = []
+                for whl_pkg_dir in pkg_dirs:
+                    layer_builder = package_builder.PackageLayerBuilder(
+                        self._ctx, self._descriptor_files, whl_pkg_dir,
+                        interpreter_builder)
+                    with ftl_util.Timing("building pkg layer"):
+                        layer_builder.BuildLayer()
+                    req_txt_imgs.append(layer_builder)
+
+                with ftl_util.Timing("stitching lyrs into req.txt image"):
+                    req_txt_image = self.AppendLayersIntoImage(req_txt_imgs)
 
-            with ftl_util.Timing("installing pip packages"):
-                self._pip_install(pkg_descriptor)
+                req_txt_builder.SetImage(req_txt_image)
+                with ftl_util.Timing("uploading req.txt image"):
+                    self._cache.Set(req_txt_builder.GetCacheKey(),
+                                    req_txt_builder.GetImage())
 
-            with ftl_util.Timing("resolving whl paths"):
-                whls = self._resolve_whls()
-                pkg_dirs = [self._whl_to_fslayer(whl) for whl in whls]
-            for whl_pkg_dir in pkg_dirs:
-                layer_builder = package_builder.PackageLayerBuilder(
-                    self._ctx, self._descriptor_files, whl_pkg_dir,
-                    interpreter_builder)
-                cached_pkg_img = None
                 if self._args.cache:
-                    with ftl_util.Timing("checking cached pkg layer"):
-                        key = layer_builder.GetCacheKey()
-                        cached_pkg_img = self._cache.Get(key)
-                if cached_pkg_img is not None:
-                    layer_builder.SetImage(cached_pkg_img)
-                else:
-                    with ftl_util.Timing("building pkg layer"):
-                        layer_builder.BuildLayer()
-                    if self._args.cache:
-                        with ftl_util.Timing("uploading pkg layer"):
-                            self._cache.Set(layer_builder.GetCacheKey(),
-                                            layer_builder.GetImage())
-                lyr_imgs.append(layer_builder)
+                    with ftl_util.Timing("uploading req.txt layer"):
```

Comment:
this is really an image, not a layer, right?

Reply:
done